File: python-impl/ol_by_tltll.py
from ol_by_tll import OLByTLL
from doubly_linked_list import DoublyLinkedList, TaggedNode, TaggedNodeWithRep
import math

class OLByTLTLL(OLByTLL):
    def __init__(self, x):
        self.reps = DoublyLinkedList()
        self.value_to_node = dict()
        
        x_node = TaggedNodeWithRep(x, 0)
        self.value_to_node[x] = x_node
        
        x_rep = TaggedNode(DoublyLinkedList(), 0)
        x_rep.value.insert(x_node)
        x_node.rep = x_rep
        
        self.reps.insert(x_rep)
        
        self.n = 1
        self.N = 2
    
    def insert(self, x, y):
        x_node = self.value_to_node[x]
        
        if (self.violates_invariant_1(self.n + 1, self.N)):
            self.rebuild()
        
        x_sublist = x_node.rep.value
        if (self.violates_invariant_2(len(x_sublist) + 1, self.sublist_N())):
            self.split_sublist(x_sublist)
            x_sublist = x_node.rep.value
        
        sublist_u = self.calculate_u(self.sublist_N())
        if (not self.available_tag_after(x_node, sublist_u)):
            self.relabel(x_node, len(x_sublist), sublist_u)
        
        y_node = TaggedNodeWithRep(y, self.get_new_tag_after(x_node, sublist_u))
        self.value_to_node[y] = y_node
        x_sublist.insertAfter(x_node, y_node)
        y_node.rep = x_node.rep
        self.n += 1
    
    def delete(self, x):
        x_node = self.value_to_node.pop(x)
        x_sublist = x_node.rep.value
        
        if (self.violates_invariant_1(self.n - 1, self.N)):
            self.rebuild()
        
        if (self.violates_invariant_2(len(x_sublist) - 1, self.sublist_N())):
            self.reps.remove(x_node.rep)
        
        x_sublist.remove(x_node)
        self.n -= 1
    
    def order(self, x, y):
        x_node = self.value_to_node[x]
        y_node = self.value_to_node[y]
        
        if (x_node.rep.tag == y_node.rep.tag):
            return x_node.tag < y_node.tag
        return x_node.rep.tag < y_node.rep.tag
    
    def sublist_N(self):
        if (self.N == 1):
            return 1
        return int(math.log2(self.N))
    
    def rep_N(self):
        if (self.N == 1):
            return 1
        return int(self.N / self.sublist_N())
    
    def violates_invariant_2(self, n_sublist, N_sublist):
        return not (0 < n_sublist <= 2 * N_sublist)
    
    def split_sublist(self, sublist):
        first_half_rep = sublist.head.rep
        first_half_rep.value = DoublyLinkedList()
        second_half_rep = TaggedNode(DoublyLinkedList(), 0)
        
        half_count = len(sublist) // 2 # should equal sublist_N
        
        node = sublist.head
        for i in range(0, len(sublist)):
            next_node = node.next
            rep = first_half_rep if i < half_count else second_half_rep
            rep.value.append(node)
            node.rep = rep
            
            node = next_node
        
        sublist_u = self.calculate_u(self.sublist_N())
        self.assign_new_tags(first_half_rep.value.head, half_count, 0, sublist_u)
        self.assign_new_tags(second_half_rep.value.head, half_count, 0, sublist_u)  
        
        if not self.available_tag_after(first_half_rep, self.calculate_u(self.rep_N())):
            self.relabel(first_half_rep, len(self.reps), self.calculate_u(self.rep_N()))
        
        second_half_rep.tag = self.get_new_tag_after(first_half_rep, self.calculate_u(self.rep_N()))
        self.reps.insertAfter(first_half_rep, second_half_rep)
    
    def rebuild(self):
        print("rebuild")
        self.N = self.n
        sublist_N = self.sublist_N()
        rep_N = self.rep_N()
        
        old_rep = self.reps.head
        node = old_rep.value.head
        self.reps = DoublyLinkedList()        
        
        while node is not None:
            new_rep = TaggedNode(DoublyLinkedList())
            
            for i in range(0, sublist_N):                
                next_node = node.next
                new_rep.value.append(node)
                node.rep = new_rep
                node = next_node
                
                if node is None:
                    old_rep = old_rep.next
                    if old_rep is None:
                        break
                    node = old_rep.value.head
            
            self.assign_new_tags(new_rep.value.head, len(new_rep.value), 0, self.calculate_u(sublist_N))
            self.reps.append(new_rep)
        
        self.assign_new_tags(self.reps.head, len(self.reps), 0, self.calculate_u(rep_N))
    
    def __repr__(self):
        sublist_N = self.sublist_N()
        rep_N = self.rep_N()
        sublist_u = self.calculate_u(sublist_N)
        rep_u = self.calculate_u(rep_N)
        text = "OLByTLTLL: N:" + str(self.N) + " n:" + str(self.n) + " subl_N:" + str(sublist_N) + " subl_u:" + str(sublist_u) + " rep_N:" + str(rep_N) + " rep_u:" + str(rep_u)
        for rep in self.reps:
            text += " ->  ((" + "[" + str(rep.tag) + "] n:" + str(len(rep.value)) + "))"
            for node in rep.value:
                text += " -> " + "[" + str(node.tag) + "]" + str(node.value)
        return text
        
        

File: js-vis/js-impl/olByTltll.js
class OlByTltll {
    constructor(x) {
        ol = this;
        visualisation.clearMessages();
        visualisation.addSequence();
        visualisation.logMessage("Initialize(" + x + ")", "blue", false);
        visualisation.addMessageIndent(1);

        this.reps = new DoublyLinkedList();
        this.valueToNode = new Map();
        let xNode = new TaggedNodeWithRep(x, 0);
        this.valueToNode.set(x, xNode);
        let xRep = new TaggedNode(new DoublyLinkedList(), 0);
        xRep.value.insert(xNode);
        xNode.rep = xRep;
        this.reps.insert(xRep);
        this.n = 1;
        this.N = 2;

        visualisation.refresh(false);
        visualisation.logMessage("initialized new ordered list with value " + x, 'green', false);
        visualisation.addMessageIndent(-1);
        visualisation.process();
    }

    insert(x, y) {
        visualisation.addSequence();
        visualisation.logMessage("Insert(" + x + ", " + y + ")","blue", false);
        visualisation.addMessageIndent(1);

        let xNode = this.valueToNode.get(x);
        visualisation.highlight(node => node.data().value === x, "blue", true, false);

        if (OlUtils.violatesInvariant1(this.n + 1, this.N)) {
            visualisation.logMessage("rebuild (n = " + this.n + ", N = " + this.N + ", n + 1 > 2 * N, violating invariant 1)", "blue", true);
            visualisation.addMessageIndent(1);
            this.rebuild();
            visualisation.addMessageIndent(-1);
            visualisation.highlight(node => node.data().value === x, "blue",true,  false);
        }

        let xSublist = xNode.rep.value;

        if (OlUtils.violatesInvariant2(xSublist.length + 1, this.sublistN())) {
            visualisation.logMessage("split sublist in half (n_s = " + xSublist.length + ", N_s = " + this.sublistN() + ", n_s + 1 > 2 * N_s, violating invariant 2)", "blue", true);
            visualisation.addMessageIndent(1);
            this.splitSublist(xSublist);
            xSublist = xNode.rep.value;
            visualisation.addMessageIndent(-1);
            visualisation.highlight(node => node.data().value === x,"blue",  true, false);
        }

        let sublistU = OlUtils.calculateU(this.sublistN());

        if (!OlUtils.availableTagAfter(xNode, sublistU)) {
            visualisation.logMessage("relabel (no available tag after " + x, "blue", true);
            visualisation.addMessageIndent(1);
            OlUtils.relabel(xNode, xSublist.length, sublistU, xNode.rep.tag);
            visualisation.addMessageIndent(-1);
            visualisation.highlight(node => node.data().value === x, "blue",true,  true);
        }

        let yNode = new TaggedNodeWithRep(y, OlUtils.getNewTagAfter(xNode, sublistU));
        this.valueToNode.set(y, yNode);
        xSublist.insertAfter(xNode, yNode);
        yNode.rep = xNode.rep;
        this.n += 1;

        visualisation.refresh(true);
        visualisation.highlight(node => node.data().value === y, "green", false, false);
        visualisation.logMessage("inserted " + y + " after " + x , "green", false);
        visualisation.addMessageIndent(-1);
        visualisation.process();
    }

    delete(x) {
        visualisation.addSequence();
        visualisation.logMessage("Delete(" + x + ")", "blue", false);
        visualisation.addMessageIndent(1);
        visualisation.highlight((node => node.data().value === x), "red", true, false);
        let xNode = this.valueToNode.get(x);

        if (OlUtils.violatesInvariant1(this.n - 1, this.N)) {
            visualisation.logMessage("rebuild (n = " + this.n + ", N = " + this.N + ", N / 2 < n - 1, violating invariant 1)", "blue", true);
            visualisation.addMessageIndent(1);
            this.rebuild();
            visualisation.addMessageIndent(-1);
            visualisation.highlight((node => node.data().value === x), "red", false, false);
        }

        let xSublist = xNode.rep.value;

        if (OlUtils.violatesInvariant2(xSublist.length - 1, this.sublistN())) {
            visualisation.logMessage("also delete sublist rep (sublist is empty, violating invariant 2)", "blue", true);
            visualisation.highlight((node => node.data().tag === xNode.rep.tag && node.data().value === "sublistRep"), "red", false, false);
            this.reps.remove(xNode.rep);
        }

        xSublist.remove(xNode);
        this.valueToNode.delete(x);
        this.n -= 1;

        visualisation.refresh(true);
        visualisation.logMessage("deleted " + x, "green", false);
        visualisation.addMessageIndent(-1);
        visualisation.process();
    }

    order(x, y) {
        visualisation.addSequence();
        visualisation.logMessage("Order(" + x + ", " + y + ")", "blue", false);
        visualisation.addMessageIndent(1);
        visualisation.highlight((node => node.data().value === x), "blue", true, false);
        visualisation.highlight((node => node.data().value === y), "blue", false, false);

        let xNode = this.valueToNode.get(x);
        let yNode = this.valueToNode.get(y);
        let result = false;

        if (xNode.rep.tag === yNode.rep.tag) {
            visualisation.logMessage(x + ".rep.tag == " + y + ".rep.tag", "blue", true);
            visualisation.addMessageIndent(1);
            result = xNode.tag < yNode.tag;
            visualisation.logMessage(x + ".tag " + (result ? "<" : ">") + " " + y + ".tag", "blue", true);
            visualisation.addMessageIndent(-1);
        }

        else {
            result = xNode.rep.tag < yNode.rep.tag;
            visualisation.logMessage(x + ".rep.tag " + (result ? "<" : ">") + " " + y + ".rep.tag", "blue", true);
        }

        visualisation.logMessage("order returned: " + result + ", " + x + (result? " < " : " > ") + y, "green", true);
        visualisation.addMessageIndent(-1);
        visualisation.process();
        return result;
    }

    sublistN() {
        if (this.N === 1) {
            return 1;
        }

        return Math.ceil(Math.log2(this.N));
    }

    repN() {
        if (this.N === 1) {
            return 1;
        }

        return Math.ceil(this.N / this.sublistN());
    }

    splitSublist(sublist) {
        let firstHalfRep = sublist.head.rep;

        if (!OlUtils.availableTagAfter(firstHalfRep, OlUtils.calculateU(this.repN()))) {
            visualisation.logMessage("relabel reps (no available tag for new sublist rep)", "blue", true);
            OlUtils.relabel(firstHalfRep, this.reps.length, OlUtils.calculateU(this.repN()));
            // visualisation.refresh(true);
        }

        firstHalfRep.value = new DoublyLinkedList();
        let secondHalfRep = new TaggedNode(new DoublyLinkedList(), 0);
        let halfCount = Math.floor(sublist.length / 2);
        let node = sublist.head;

        for (let i = 0; i < sublist.length; i++) {
            let nextNode = node.next;
            let rep = i < halfCount ? firstHalfRep : secondHalfRep;
            rep.value.append(node);
            node.rep = rep;
            node = nextNode;
        }

        let sublistU = OlUtils.calculateU(this.sublistN());
        OlUtils.assignNewTags(firstHalfRep.value.head, halfCount, 0, sublistU);
        OlUtils.assignNewTags(secondHalfRep.value.head, halfCount, 0, sublistU);

        secondHalfRep.tag = OlUtils.getNewTagAfter(firstHalfRep, OlUtils.calculateU(this.repN()));
        this.reps.insertAfter(firstHalfRep, secondHalfRep);

        visualisation.logMessage("split sublist into 2", "blue", true);
        visualisation.refresh(false);
    }

    rebuild() {
        this.N = this.n;
        visualisation.logMessage("set new N = n =" + this.N, "blue", true);
        visualisation.logMessage("set new N_s = log(N) = " + this.sublistN(), "blue", false);
        visualisation.logMessage("set new u_s = 2 * N_s ^ 2 = " + OlUtils.calculateU(this.sublistN()), "blue", false);
        visualisation.logMessage("set new N_r = N / N_s = " + this.repN(), "blue", false);
        visualisation.logMessage("set new u_r = 2 * N_r ^ 2 = " + OlUtils.calculateU(this.repN()), "blue", false);

        let sublistN = this.sublistN();
        let repN = this.repN();
        let oldRep = this.reps.head;
        let node = oldRep.value.head;
        this.reps = new DoublyLinkedList();

        while (node) {
            let newRep = new TaggedNode(new DoublyLinkedList());

            for (let i = 0; i < sublistN; i++) {
                let nextNode = node.next;
                newRep.value.append(node);
                node.rep = newRep;
                node = nextNode;

                if (!node) {
                    oldRep = oldRep.next;

                    if (!oldRep) {
                        break;
                    }

                    node = oldRep.value.head;
                }
            }

            OlUtils.assignNewTags(newRep.value.head, newRep.value.length, 0, OlUtils.calculateU(sublistN));
            this.reps.append(newRep);
        }

        OlUtils.assignNewTags(this.reps.head, this.reps.length, 0, OlUtils.calculateU(repN));

        visualisation.refresh(false);
        visualisation.logMessage("evenly distributed nodes into " + this.reps.length + " sublists", "blue", false);
        visualisation.logMessage("set new (evenly distributed) tags for all reps", "blue", false);
        visualisation.logMessage("set new (evenly distributed) tags for all sublist nodes", "blue", false);
    }

    toString() {
        let text = "OlByTltll ";
        for (let rep of this.reps) {
            text += rep.toString() + "  ";
        }

        return text;
    }

    contains(x) {
        return this.valueToNode.has(x);
    }

    getNumberOfElements() {
        return this.n;
    }

    getProperties() {
        let properties = [
            "Ordered list by tagged linked list",
            "n: " + this.n,
            "N: " + this.N,
            "N_s: " + this.sublistN(),
            "u_s: " + OlUtils.calculateU(this.sublistN()),
            "N_r: " + this.repN(),
            "u_r: " + OlUtils.calculateU(this.repN()),
        ]

        return properties;
    }

    toGraph(treeViewOn) {
        if (treeViewOn) {
            return OlToGraph.toGraphTltllTree(this.reps, OlUtils.calculateU(this.repN()), OlUtils.calculateU(this.sublistN()));
        }
        return OlToGraph.toGraphTltll(this.reps);
    }
}
